### PR TITLE
Merge pr 331, with some fixups

### PR DESCRIPTION
**Pull request by peihe**
Additional fixups:
1. Uses the default MAX_RPC_ATTEMPTS, and reduced number of the parameters.
2. Uses errorExtractor
3. Removed testExportJobDoesNotExist, since it is same as testReadFromTable.
4. Adds getJob() unit tests.

### CODE REVIEW

**Comment by peihe** [comment]
R: @dhalperi

**Comment by dhalperi** [comment]
CLA confirmed for Google and Spotify.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java [hunk 1]
```diff
@@ -244,6 +244,46 @@ public JobStatistics dryRunQuery(String projectId, String query)
...
+    @VisibleForTesting
+    public Job getJob(JobReference jobRef, Sleeper sleeper, BackOff backoff)
+        throws IOException, InterruptedException {
+      Exception lastException;
+      do {
+        try {
+          return client.jobs().get(jobRef.getProjectId(), jobRef.getJobId()).execute();
+        } catch (GoogleJsonResponseException e) {
+          if (errorExtractor.itemNotFound(e)) {
+            LOG.info("Job {} does not exists.", jobRef.getJobId());
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java]: A little more detail: "No BigQuery job with job id {} found".

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java [hunk 2]
```diff
@@ -244,6 +244,46 @@ public JobStatistics dryRunQuery(String projectId, String query)
...
+            return null;
+          }
+          LOG.warn("Ignore the error and retry getting the job.", e);
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java]: ("Ignoring error encountered while trying to query the status of BigQuery job {}", jobId, e)

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java [hunk 3]
```diff
@@ -244,6 +244,46 @@ public JobStatistics dryRunQuery(String projectId, String query)
...
+          lastException = e;
+        } catch (IOException e) {
+          LOG.warn("Ignore the error and retry getting the job.", e);
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java]: similar

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java [hunk 4]
```diff
@@ -244,6 +244,46 @@ public JobStatistics dryRunQuery(String projectId, String query)
...
+          lastException = e;
+        }
+      } while (nextBackOff(sleeper, backoff));
+      throw new IOException(
+          String.format(
+              "Unable to find job: %s, aborting after %d retries.",
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java]: BigQuery job

**Comment by dhalperi** [comment]
LGTM, with minor cleanups

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java (hunk 1 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java]: done

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java (hunk 2 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java]: done

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java (hunk 3 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java]: done

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java (hunk 4 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryServicesImpl.java]: done

**Comment by peihe** [comment]
Updated the error messages.
PTAL

**Comment by dhalperi** [comment]
Does not compile? PTAL

**Comment by peihe** [comment]
Fixed

**Comment by dhalperi** [comment]
Thanks.

Does this need to be forward-ported to Apache Beam?

**Pull request merged**